### DISCUSSION TITLE
Unable to use HTTPS with Wiki.js and Docker

### DISCUSSION THREAD

**Discussion by 2bitsgmbh** [requarks/wiki#4635] [Help / Questions] [answered]
Hi all,

I followed this Tutorial: https://docs.requarks.io/install/ubuntu

**Host Info:**
OS: Ubuntu 20.04.3 LTS with Docker
Wiki.js: 2.5.219
Node.js: 14.17.6
PostgreSQL: 11.11

The first part is working and wiki.js is up and running with http on port 80.
Now I tried the part "Automatic HTTPS with Let's Encrypt (optional)" and getting this error:

```
Loading configuration from /wiki/config.yml... OK
DB_PASS_FILE is defined. Will use secret from file.
2021-10-26T16:14:56.958Z [MASTER] info: =======================================
2021-10-26T16:14:56.961Z [MASTER] info: = Wiki.js 2.5.219 =====================
2021-10-26T16:14:56.961Z [MASTER] info: =======================================
2021-10-26T16:14:56.961Z [MASTER] info: Initializing...
2021-10-26T16:14:57.804Z [MASTER] info: Using database driver pg for postgres [ OK ]
2021-10-26T16:14:57.811Z [MASTER] info: Connecting to database...
2021-10-26T16:14:57.843Z [MASTER] info: Database Connection Successful [ OK ]
2021-10-26T16:14:58.770Z [MASTER] info: Loading GraphQL Schema...
2021-10-26T16:14:59.836Z [MASTER] info: GraphQL Schema: [ OK ]
2021-10-26T16:15:00.225Z [MASTER] info: HTTP Server on port: [ 3000 ]
2021-10-26T16:15:00.277Z [MASTER] info: (LETSENCRYPT) Initializing Let's Encrypt client...
2021-10-26T16:15:00.310Z [MASTER] info: HTTP Server: [ RUNNING ]
2021-10-26T16:15:00.752Z [MASTER] info: (LETSENCRYPT) Generating server keypairs...
2021-10-26T16:15:01.066Z [MASTER] info: (LETSENCRYPT) Server keypairs generated successfully [ OK ]
2021-10-26T16:15:01.067Z [MASTER] info: (LETSENCRYPT) Generating certificate signing request (CSR)...
2021-10-26T16:15:01.077Z [MASTER] info: (LETSENCRYPT) CSR generated successfully [ OK ]
2021-10-26T16:15:01.078Z [MASTER] info: (LETSENCRYPT) Requesting certificate from Let's Encrypt...
2021-10-26T16:15:01.087Z [MASTER] info: (LETSENCRYPT) Setting HTTP challenge for wiki.2bits.de: [ READY ]
2021-10-26T16:15:01.088Z [MASTER] info: (LETSENCRYPT) Waiting for challenge to complete...
2021-10-26T16:15:06.152Z [MASTER] info: (LETSENCRYPT) Removing HTTP challenge: [ OK ]
2021-10-26T16:15:06.153Z [MASTER] info: (LETSENCRYPT) Removing HTTP challenge: [ OK ]
2021-10-26T16:15:06.154Z [MASTER] warn: (LETSENCRYPT) Error: Error: Failed HTTP-01 Pre-Flight / Dry Run.
curl 'http://wiki.2bits.de/.well-known/acme-challenge/test-1f448afd3f837010e15aeab43a354f63-0'
Expected: 'test-1f448afd3f837010e15aeab43a354f63-0.oBnhst2ibLwKe04fb0esuY2YLdCWI_JRQzE2Wanmf1c'
Got: '<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
<html><head>
<title>404 Not Found</title>
</head><body>
<h1>Not Found</h1>
<p>The requested URL /.well-known/acme-challenge/test-1f448afd3f837010e15aeab43a354f63-0 was not found on this server.</p>
</body></html>
'
See https://git.rootprojects.org/root/acme.js/issues/4
2021-10-26T16:15:06.154Z [MASTER] error: Error: Failed HTTP-01 Pre-Flight / Dry Run.
curl 'http://wiki.2bits.de/.well-known/acme-challenge/test-1f448afd3f837010e15aeab43a354f63-0'
Expected: 'test-1f448afd3f837010e15aeab43a354f63-0.oBnhst2ibLwKe04fb0esuY2YLdCWI_JRQzE2Wanmf1c'
Got: '<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
<html><head>
<title>404 Not Found</title>
</head><body>
<h1>Not Found</h1>
<p>The requested URL /.well-known/acme-challenge/test-1f448afd3f837010e15aeab43a354f63-0 was not found on this server.</p>
</body></html>
'
See https://git.rootprojects.org/root/acme.js/issues/4
```

Can somebody please help?

**Comment by ieaster1** [comment]
Could you provide either the Dockerfile or command you're using to start up the container?  Are you using any proxies in front of the container for access?

**Reply by 2bitsgmbh** [reply]
I'm using this command from the tutorial

`docker create --name=wiki -e LETSENCRYPT_DOMAIN=wiki.2bits.de -e LETSENCRYPT_EMAIL=server@2bits.de -e SSL_ACTIVE=1 -e DB_TYPE=postgres -e DB_HOST=db -e DB_PORT=5432 -e DB_PASS_FILE=/etc/wiki/.db-secret -v /etc/wiki/.db-secret:/etc/wiki/.db-secret:ro -e DB_USER=wiki -e DB_NAME=wiki -e UPGRADE_COMPANION=1 --restart=unless-stopped -h wiki --network=wikinet -p 80:3000 -p 443:3443 requarks/wiki:2`

I blindly followed the tutotial. I have not set up a proxy.
The server with Wiki.js got its own public IP and this was linked with an A record in the DNS with the subdomain.
The firewall/router let ports 443 and 80 through to the server. It is only routed from the public IP to the internal IP of the server.


This command without Let's Encrypt and just using HTTP on port 80 is working:
`docker create --name=wiki -e DB_TYPE=postgres -e DB_HOST=db -e DB_PORT=5432 -e DB_PASS_FILE=/etc/wiki/.db-secret -v /etc/wiki/.db-secret:/etc/wiki/.db-secret:ro -e DB_USER=wiki -e DB_NAME=wiki -e UPGRADE_COMPANION=1 --restart=unless-stopped -h wiki --network=wikinet -p 80:3000 -p 443:3443 requarks/wiki:2`

**Comment by 2bitsgmbh** [comment]
Is there anyone who can help?

**Comment by ieaster1** [comment, marked as answer]
@2bitsgmbh I just ran a test environment following the guide and everything completed without any issues or extra tweaking.

My suggestion, purge your current environment and start fresh if you can.  That means, remove containers, network, and volumes.

There also isn't a need to create a container without the Let's Encrypt config. I would consider this the correct process if you want HTTPS for this step:
```
# Create installation directory for Wiki.js
mkdir -p /etc/wiki

# Generate DB secret
openssl rand -base64 32 > /etc/wiki/.db-secret

# Create internal docker network
docker network create wikinet

# Create data volume for PostgreSQL
docker volume create pgdata

# Create the containers
docker create --name=db -e POSTGRES_DB=wiki -e POSTGRES_USER=wiki -e POSTGRES_PASSWORD_FILE=/etc/wiki/.db-secret -v /etc/wiki/.db-secret:/etc/wiki/.db-secret:ro -v pgdata:/var/lib/postgresql/data --restart=unless-stopped -h db --network=wikinet postgres:11
docker create --name=wiki -e LETSENCRYPT_DOMAIN=wiki.example.com -e LETSENCRYPT_EMAIL=admin@example.com -e SSL_ACTIVE=1 -e DB_TYPE=postgres -e DB_HOST=db -e DB_PORT=5432 -e DB_PASS_FILE=/etc/wiki/.db-secret -v /etc/wiki/.db-secret:/etc/wiki/.db-secret:ro -e DB_USER=wiki -e DB_NAME=wiki -e UPGRADE_COMPANION=1 --restart=unless-stopped -h wiki --network=wikinet -p 80:3000 -p 443:3443 requarks/wiki:2
docker create --name=wiki-update-companion -v /var/run/docker.sock:/var/run/docker.sock:ro --restart=unless-stopped -h wiki-update-companion --network=wikinet requarks/wiki-update-companion:latest
```

If that still isn't working for you it sounds like there could be something else going on.  Maybe a local DNS issues or you may have DNSSEC enabled and need to use a CAA Record instead of an A record for your domain name.

**Reply by 2bitsgmbh** [reply]
I'm sure you know this too. It had to go very quickly and now it's been used internally for some time.
If I have to set it up again, then I have to try to migrate everything.

I will test this on a virtual machine directly with Let's Encrypt and if it doesn't work there either, I will have to look for the error somewhere else.

Thanks so far for your help!

**Reply by ieaster1** [reply]
Absolutely, the every present "done yesterday."

As long as you keep the db volume your data will be safe.  I'd agree to test it on a different platform then production though.

If you created the container without SSL and then switched over, that could be causing your funkiness but that is only a theory.

**Comment by 2bitsgmbh** [comment]
I have now started completely from scratch on another machine directly with Let's Encrypt and that worked.

Now I have to figure out how to get the data from the old wiki into the new one.
@ieaster1 do you have a hint?

**Reply by ieaster1** [reply]
This [stack](https://stackoverflow.com/questions/1237725/copying-postgresql-database-to-another-server) should help you out of copying the data from one postgres instance to another.

Really, you just want to dump the DB and import it.

**Reply by 2bitsgmbh** [reply]
I found this one
[https://docs.requarks.io/install/transfer](url)
and it worked well.

Thank you for your help and support!
